### PR TITLE
1.21

### PR DESCRIPTION
**Pull request by dima-dencep**
Now this reworks the neoforge code a bit and also removes your implementation of HolderLookup$Provider

### CODE REVIEW

File: minecraft/forge/src/main/java/dev/kosmx/playerAnim/forge/ForgeClientEvent.java [hunk 1]
```diff
@@ -5,31 +5,31 @@
...
+
+    public ForgeClientEvent(IEventBus bus) {
+        bus.addListener(this::resourceLoadingListener);
 
-    @SubscribeEvent
-    public static void clientSetup(FMLClientSetupEvent event) {
         if (Helper.isBendEnabled() && ModList.get().isLoaded("skinlayers3d")) {
             try {
-                SkinLayersTransformer.init(ForgeModInterface.LOGGER);
+                SkinLayersTransformer.init(LOGGER);
```
Comment (KosmX) [inline minecraft/forge/src/main/java/dev/kosmx/playerAnim/forge/ForgeClientEvent.java]: This is intentionally setup event and not class ctor

File: minecraft/forge/src/main/java/dev/kosmx/playerAnim/forge/ForgeClientEvent.java [hunk 2]
```diff
@@ -5,31 +5,31 @@
...
-                ForgeModInterface.LOGGER.error("Failed to initialize 3D skin layers compat: " + e.getMessage());
+                LOGGER.error("Failed to initialize 3D skin layers compat: " + e.getMessage());
             }
         }
     }
 
-    @SubscribeEvent
-    public static void resourceLoadingListener(@NotNull RegisterClientReloadListenersEvent event) {
-        event.registerReloadListener((ResourceManagerReloadListener) manager -> PlayerAnimationRegistry.resourceLoaderCallback(manager, ForgeModInterface.LOGGER));
+    public void resourceLoadingListener(@NotNull RegisterClientReloadListenersEvent event) {
```
Comment (KosmX) [inline minecraft/forge/src/main/java/dev/kosmx/playerAnim/forge/ForgeClientEvent.java:36]: does this work if not static?
forge event bus is broken...

**Comment by KosmX** [comment]
There are some small things that makes sense, but mostly no.

**Comment by KosmX** [comment]
I understand if you're trying to help, but checking whether the already existing code works can help more.

File: minecraft/forge/src/main/java/dev/kosmx/playerAnim/forge/ForgeClientEvent.java (hunk 2 above)
Comment (dima-dencep) [reply minecraft/forge/src/main/java/dev/kosmx/playerAnim/forge/ForgeClientEvent.java:36]: There's a subscriber in the ctor, everything's fine.

File: minecraft/forge/src/main/java/dev/kosmx/playerAnim/forge/ForgeClientEvent.java (hunk 1 above)
Comment (dima-dencep) [reply minecraft/forge/src/main/java/dev/kosmx/playerAnim/forge/ForgeClientEvent.java]: why

**Comment by dima-dencep** [comment]
If you're happy with what's here now, that's too bad, but I don't see any solutions other than creating your own fork.

File: minecraft/forge/src/main/java/dev/kosmx/playerAnim/forge/ForgeClientEvent.java (hunk 1 above)
Comment (KosmX) [reply minecraft/forge/src/main/java/dev/kosmx/playerAnim/forge/ForgeClientEvent.java]: mod class ctor will be invoked way earlier than seteup event.
doing some event subscribe from ctor is okay, but it is adviced, not to do any real initialization/etc from the ctor. The mod setup event is for that

File: minecraft/common/src/main/java/dev/kosmx/playerAnim/minecraftApi/PlayerAnimationRegistry.java [hunk 3]
```diff
@@ -122,17 +119,7 @@ public static void resourceLoaderCallback(@NotNull ResourceManager manager, Logg
...
-                public @NotNull Stream<ResourceKey<? extends Registry<?>>> listRegistries() {
-                    return Stream.empty();
-                }
-
-                @Override
-                public <T> @NotNull Optional<HolderLookup.RegistryLookup<T>> lookup(@NotNull ResourceKey<? extends Registry<? extends T>> resourceKey) {
-                    return Optional.empty();
-                }
-            });
+            var component = Component.Serializer.fromJson(arg, RegistryAccess.EMPTY);
```
Comment (KosmX) [inline minecraft/common/src/main/java/dev/kosmx/playerAnim/minecraftApi/PlayerAnimationRegistry.java]: yup, this one is okay.

**Comment by KosmX** [comment]
I don't say don't open a pull-request, and yes, there are some stuff I should clean up.

But There are a lot of stuff that are there for a reason. Minecraft modding is not always clean code, especially if Forge is involved

File: minecraft/forge/src/main/java/dev/kosmx/playerAnim/forge/ForgeClientEvent.java (hunk 1 above)
Comment (KosmX) [reply minecraft/forge/src/main/java/dev/kosmx/playerAnim/forge/ForgeClientEvent.java]: The other reason why you don't delete the ForgeModInterface.java:
This library contains platform independent code, modders must be able to load this on server-side.
If the ForgeModInterface is client-only, Forge will simply fail to load it on server.

And the setup event must not run on server

File: minecraft/forge/src/main/java/dev/kosmx/playerAnim/forge/ForgeClientEvent.java
Comment (KosmX) [inline minecraft/forge/src/main/java/dev/kosmx/playerAnim/forge/ForgeClientEvent.java:1]: Please keep the setup event and use that for setup stuff.

File: minecraft/forge/src/main/java/dev/kosmx/playerAnim/impl/neoforge/HelperImpl.java
Comment (KosmX) [inline minecraft/forge/src/main/java/dev/kosmx/playerAnim/impl/neoforge/HelperImpl.java:1]: I'll verify that, but looks good

**Review by KosmX — CHANGES_REQUESTED** [review]

**Review by KosmX — APPROVED** [review]

**Pull request merged**